File: src/modifiers/onTapGesture.ts
import { createModifier } from './createModifier';

export interface OnTapGestureParams {
  /// Number of taps required to fire. Defaults to 1.
  count?: number;
  /// Fired when the tap gesture completes.
  handler: () => void;
}

/// SwiftUI `.onTapGesture(count:perform:)`. Fires `handler` after `count`
/// taps land on the view.
///
/// ```tsx
/// <Text modifiers={[onTapGesture(() => setCount((c) => c + 1))]} />
/// ```
export function onTapGesture(
  handler: () => void
): ReturnType<typeof createModifier>;
export function onTapGesture(
  params: OnTapGestureParams
): ReturnType<typeof createModifier>;
export function onTapGesture(a: OnTapGestureParams | (() => void)) {
  if (typeof a === 'function') {
    return createModifier('onTapGesture', { count: 1, handler: a });
  }
  return createModifier('onTapGesture', { count: a.count ?? 1, handler: a.handler });
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/OnTapGesture.swift
import SwiftUI

/// SwiftUI `.onTapGesture(count:perform:)`. Fires the JS `handler` after
/// `count` taps (default 1).
enum RNWOnTapGestureModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("onTapGesture") { view, params, ctx in
            let count = params.int("count") ?? 1
            let handlerId = params.int("handler")
            return AnyView(view.onTapGesture(count: count) {
                ctx.fire(handlerId)
            })
        }
    }
}
